### PR TITLE
149: execute payment only if there are no successful transactions of type "Charge"

### PR DESCRIPTION
**Pull request by JudeNiroshan**
Issue : https://github.com/commercetools/commercetools-paypal-plus-integration/issues/149

### CODE REVIEW

File: src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java [hunk 1]
```diff
@@ -201,12 +210,56 @@ public void shouldCreateChargeTransaction() {
...
+        final String ctpPaymentId = updatedPayment.getId();
+
+        executeBlocking(paymentHandler.createPayment(ctpPaymentId));
+
+        PaymentHandleResponse response = executeBlocking(paymentHandler
+                .executePayment(updatedPayment.getInterfaceId(), "testPayerId"));
+
+        io.sphere.sdk.payments.Payment ctpPayment = executeBlocking(sphereClient
+                .execute(PaymentByIdGet.of(ctpPaymentId)));
+        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED.value());
```
Comment (lojzatran) [inline src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java]: How do you know from this test that it was not created when running `executePayment` method?

File: src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java [hunk 2]
```diff
@@ -201,12 +210,56 @@ public void shouldCreateChargeTransaction() {
         assertThat(chargeTxn.getState()).isEqualTo(SUCCESS);
     }
 
+    @Test
+    public void whenCtpPaymentIsAlreadyMarkedAsSuccess_shoudNotCallCreateTransactionInPaypalPlus() {
+
```
Comment (lojzatran) [inline src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java]: Remove this line

File: src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java [hunk 3]
```diff
@@ -201,12 +210,56 @@ public void shouldCreateChargeTransaction() {
...
+                .execute(PaymentByIdGet.of(ctpPaymentId)));
+        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED.value());
+        assertThat(ctpPayment.getInterfaceInteractions()).hasSize(1);
+    }
+
+
     @Test
     public void whenPaypalPaymentIsNotApproved_transactionShouldNotBeCreated() {
-        CtpPaymentWithCart ctpPaymentWithCart = createCartWithPayment();
-        PaymentHandler paymentHandler = paymentHandlerProvider.getPaymentHandler(MAIN_TEST_TENANT_NAME).get();
```
Comment (lojzatran) [inline src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java]: Why this blank line?

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java [hunk 4]
```diff
@@ -213,7 +213,16 @@ public PaymentHandler(@Nonnull CtpFacade ctpFacade,
...
                                         );
                                     } else {
-                                        return executePaymentAndCreateTxn(paypalPlusPaymentId, paypalPlusPayerId, ctpPayment);
+                                        boolean isAlreadyCharged = ctpPayment.getTransactions().stream()
+                                                .filter(transaction -> transaction.getType().equals(TransactionType.CHARGE))
+                                                .anyMatch(transaction -> transaction.getState().equals(SUCCESS));
+                                        if (isAlreadyCharged) {
+                                            // Should not charge for already charged transactions
+                                            return CompletableFuture.completedFuture(PaymentHandleResponse
+                                                    .ofHttpStatus(HttpStatus.CREATED));
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: yes, I already discussed about the response. He said, return the same response as it was earlier which will not break any calls for this endpoint.

File: src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java (hunk 1 above)
Comment (JudeNiroshan) [reply src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java]: It's from the number of interactions count

File: src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java (hunk 3 above)
Comment (JudeNiroshan) [reply src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java]: de5f9ee05b93c102ced24ca1c8f961aaa951a64e

File: src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java (hunk 1 above)
Comment (lojzatran) [reply src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java]: Yes, and it has 1 😄 so it was created during `executePayment`?

Comment (lojzatran) [reply src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java]: Add a verification that the number of interface interactions count didn't change after you run the execute payment.

Comment (butenkor) [reply src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java]: Why do you rely on count of interactions and not count of transactions? Change prevents obsolete transaction creation and not interaction.

Comment (JudeNiroshan) [reply src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java]: 0f6df796760637e1ed5d99a1b49dc13e078a07c8

File: src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java (hunk 2 above)
Comment (JudeNiroshan) [reply src/test/integration/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImplIT.java]: c80e2bc063492bda8fa0db75e748dae20b5e987f

**Pull request merged**